scripts/create-protection-market-curve3Crypto.ts

```
@@ -0,0 +1,127 @@
+ * @dev To deploy, you must define the CURVE_POOL environment variable, which is the address of the curve pool on the desired chain
+ */
+import hre from 'hardhat';
+import '@nomiclabs/hardhat-ethers';
+import { Contract, ContractFactory, utils } from 'ethers';
+import chalk from 'chalk';
+import { getChainId, getContractAddress, getGasPrice, logSuccess, logFailure, findLog, waitForInput, fundAccount } from '../utils/utils'; // prettier-ignore
+import comptrollerAbi from '../abi/Comptroller.json';
+
+// STEP 0: ENVIRONMENT SETUP
+const provider = hre.ethers.provider;
+const signer = new hre.ethers.Wallet(process.env.PRIVATE_KEY as string, hre.ethers.provider);
+const chainId = getChainId(hre);
+const { AddressZero } = hre.ethers.constants;
+
+// STEP 1: TRIGGER CONTRACT SETUP
+const name = 'Curve 3Crypto Trigger'; // name
+const symbol = 'CURVE-3CRYPTO-TRIG'; // symbol
+const description = "Triggers when the Curve 3Crypto pool's virtual price decreases by more than 50% between consecutive checks, or the internal balances tracked in the Curve 3Crypto pool are more than 50% lower than the true balances"; // prettier-ignore
+const platformIds = [3]; // platform ID for Curve
+const recipient = '0xSetRecipientAddressHere'; // subsidy recipient
+
+// STEP 2: TRIGGER CONTRACT DEVELOPMENT
+// For this step, see the ITrigger.sol and MockTrigger.sol examples and the corresponding documentation
+
+// STEP 3: PROTECTION MARKET DEPLOYMENT
+async function main(): Promise<void> {
+  const curvePoolAddress = String(process.env.CURVE_POOL);
+  if (!curvePoolAddress) throw new Error("Please define the 'CURVE_POOL' environment variable.");
+
+  if (!utils.isAddress(recipient)) throw new Error('\n\n**** Please set the recipient address on line 23 ****\n');
+
+  // Compile contracts to make sure we're using the latest version of the trigger contracts
+  await hre.run('compile');
+
+  const overrides = { gasPrice: await getGasPrice() };
```

Ah, we'll need to update the getGasPrice method. Right now it queries an API that is L1 only, meaning we'll overpay when deploying on other chains. I'd recommend converting getGasPrice to take an `options: { chainId: number; gasPriceConfidence: TxPriceConfidence }` object as input

Then for the logic, use the existing logic for mainnet. For other chain IDs, use `provider.getGasPrice().mul(12).div(10)`. In other words, query the node for the gas price and bump it by 20%

---

Instead of `getGasPrice` https://docs.ethers.io/v5/api/providers/provider/#Provider-getGasPrice, I was thinking we should use `getFeeData` https://docs.ethers.io/v5/api/providers/provider/#Provider-getFeeData , as we need `maxPriorityFeePerGas` and `maxFeePerGas`.. but for some reason, it returns a response object with a val for the `gasPrice` but null for `maxPriorityFeePerGas` and `maxFeePerGas`.

You're saying `maxFeePerGas = provider.getGasPrice().mul(12).div(10)` right? What would the `maxPriorityFeePerGas` be?

---

or are we able to send the transaction, just using `gasPrice` (non EIP-1559 transaction https://docs.ethers.io/v5/api/providers/provider/#Provider-getFeeData)

---

Oh cool, I didn't know about `getFeeData`! But good call about the 1559 distinction—right now only L1 and Polygon support 1559, Arbitrum and other L2s/L1s are just the legacy style gas approach. 

We have `getGasPrice` return an overrides object, and even though 1559 vs. legacy txs will have different fields, ethers is smart enough to handle them both. So I think we can still return an object that has either `{ maxFeePerGas, maxPriorityFeePerGas }` OR `{  gasPrice }`. 

So we'd pass a chain ID, and since we're not worried about polygon the above logic should still apply, i.e. we keep using the API response for L1, and for Arbitrum use either `getGasPrice` of `getFeeData`

---

And we probably should throw for unsupported chain IDs. When chain ID is 31337 (hardhat) we should check the value  of `process.env.TEST_CHAIN_FORK` to know which gas price method to use

---

updated based on the above ^, except I opted not to add the check for 31337 (hardhat) and the `TEST_CHAIN_FORK` env var, as all of our use cases for `getGasPrice` currently, we use the `getChainId` util which is based on the test chain fork url https://github.com/ScopeLift/cozy-triggers/blob/master/utils/utils.ts#L61. we can add it in the future if needed, but doesn't seem necessary at least for now